src/test/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/util/DbExtension.kt

```diff
@@ -0,0 +1,160 @@
+    val jdbcTemplate = applicationContext.getBean(JdbcTemplate::class.java)
+
+    flyway.clean()
+
+    jdbcTemplate.execute("CREATE EXTENSION postgis;")
+
+    flyway.migrate()
+  }
+
+  private fun getInitialDatabaseState(applicationContext: ApplicationContext) {
+    val entityManagerFactory = applicationContext.getBean(EntityManagerFactory::class.java)
+    val sessionFactory = entityManagerFactory.unwrap(SessionFactoryImpl::class.java)
+
+    sessionFactory.openSession().use { session ->
+      INITIAL_DB_STATE = session.metamodel.entities.associateWith {
+        val query = session.criteriaBuilder.createQuery()
+        val root = query.from(it)
+        query.select(root)
+        val results = session.createQuery(query).resultList
+
+        println("Found ${results.count()} ${it.name} instances")
+
+        results
+      }
+    }
+  }
+
+  private fun setInitialDatabaseState(applicationContext: ApplicationContext) {
+    val entityManagerFactory = applicationContext.getBean(EntityManagerFactory::class.java)
+    val sessionFactory = entityManagerFactory.unwrap(SessionFactoryImpl::class.java)
+
+    sessionFactory.openSession().use { session ->
+      val transaction = session.beginTransaction()
+
+      INITIAL_DB_STATE.forEach { (entityType, entities) ->
+        entities.forEach { session.save(it) }
+        println("Restoring ${entities.count()} ${entityType.name} instances")
+      }
+
+      session.flush()
+      transaction.commit()
+    }
+  }
+
+  private fun getDatasource(applicationContext: ApplicationContext): DataSource {
+    val configurableEnvironment = applicationContext.environment
+
+    val binder = Binder.get(configurableEnvironment)
+    val dataSourceProperties = binder
+      .bind("spring.datasource", Bindable.of(DataSourceProperties::class.java))
+      .get()
+
+    val pgSimpleDataSource = PGSimpleDataSource()
+    pgSimpleDataSource.setUrl(dataSourceProperties.url)
+    pgSimpleDataSource.user = dataSourceProperties.username
+    pgSimpleDataSource.password = dataSourceProperties.password
+    return pgSimpleDataSource
+  }
+
+  private fun cleanDatabase(dataSource: DataSource) {
+    try {
+      dataSource.connection.use { connection ->
+        connection.autoCommit = false
+        val tablesToClean = loadTablesToClean(connection)
+        cleanTablesData(tablesToClean, connection)
+        connection.commit()
+      }
+    } catch (e: SQLException) {
+      LOGGER.error(String.format("Failed to clean database due to error: \"%s\"", e.message))
+      e.printStackTrace()
+    }
+  }
+
+  private fun loadTablesToClean(connection: Connection): List<TableData> {
+    val databaseMetaData = connection.metaData
+    val resultSet = databaseMetaData.getTables(connection.catalog, null, null, arrayOf("TABLE"))
+
+    val tablesToClean = mutableListOf<TableData>()
+    while (resultSet.next()) {
+      val table = TableData(
+        schema = resultSet.getString("TABLE_SCHEM"),
```

Comment:
@danlivings-dxw - Is `TABLE_SCHEM` correct?

Reply:
The documentation for [`DatabaseMetaData.getTables`](https://docs.oracle.com/javase/7/docs/api/java/sql/DatabaseMetaData.html#getTables(java.lang.String,%20java.lang.String,%20java.lang.String,%20java.lang.String[])) specifies `"TABLE_SCHEM"` to retrieve the schema name.